# Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing Data Base

In [3]:
# Import ords_prods_merge_bdays.pkl
# Create a string of the path
# Define the path variable
path=r'/Users/douniaelyoussoufi/Achievement 4 '
# Load the ords_prods_merge.pkl file directly using os.path.join
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_bdays.pkl'))

# Step 1: Exercise 4.8

In [10]:
# Create a subset
df = ords_prods_merge[:1000000]

In [11]:
# The dimensions of this subset
df.shape

(1000000, 21)

In [12]:
# Display the first five rows of the dataframe
df.head()

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0         2       33120                  1          1   202279    prior   
1         2       28985                  2          1   202279    prior   
2         2        9327                  3          0   202279    prior   
3         2       45918                  4          1   202279    prior   
4         2       30035                  5          0   202279    prior   

   order_number  order_dow  order_hour_of_day  days_since_prior_order  ...  \
0             3          5                  9                     8.0  ...   
1             3          5                  9                     8.0  ...   
2             3          5                  9                     8.0  ...   
3             3          5                  9                     8.0  ...   
4             3          5                  9                     8.0  ...   

  Unnamed: 0           product_name aisle_id  department_id  prices  \
0    33122.0     Organic Egg Whites     86.0           16.0    11.3   
1    28987.0  Michigan Organic Kale     83.0            4.0    13.4   
2     9327.0          Garlic Powder    104.0           13.0     3.6   
3    45922.0         Coconut Butter     19.0           13.0     8.4   
4    30037.0      Natural Sweetener     17.0           13.0    13.7   

      busiest day    busiest_days    Busiest_days     busiest_day  \
0  Regularly busy  Regularly busy  Regularly busy  Regularly busy   
1  Regularly busy  Regularly busy  Regularly busy  Regularly busy   
2  Regularly busy  Regularly busy  Regularly busy  Regularly busy   
3  Regularly busy  Regularly busy  Regularly busy  Regularly busy   
4  Regularly busy  Regularly busy  Regularly busy  Regularly busy   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  

[5 rows x 21 columns]

In [13]:
# Grouping the dataframe by the 'product_name' column
df.groupby('product_name')

In [21]:
# Groups the dataframe by 'department_id' and calculates the mean of 'order_number' for each group  
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              15.629981
2.0              18.470641
3.0              17.389793
4.0              17.985849
5.0              14.412616
6.0              16.515856
7.0              17.413529
8.0              16.050344
9.0              16.083870
10.0             21.290293
11.0             16.494495
12.0             16.089124
13.0             16.787994
14.0             16.896188
15.0             16.442397
16.0             17.812191
17.0             16.081831
18.0             19.198872
19.0             17.404819
20.0             16.670273
21.0             23.252093

In [25]:
# Groups the dataframe by 'department_id' and computes the mean of 'order_number' for each group without using agg()  
df.groupby('department_id')['order_number'].mean()

department_id
1.0     15.629981
2.0     18.470641
3.0     17.389793
4.0     17.985849
5.0     14.412616
6.0     16.515856
7.0     17.413529
8.0     16.050344
9.0     16.083870
10.0    21.290293
11.0    16.494495
12.0    16.089124
13.0    16.787994
14.0    16.896188
15.0    16.442397
16.0    17.812191
17.0    16.081831
18.0    19.198872
19.0    17.404819
20.0    16.670273
21.0    23.252093
Name: order_number, dtype: float64

In [27]:
# Groups the dataframe by 'department_id' and calculates the mean, min, and max of 'order_number' for each group  
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1.0              15.629981   1  99
2.0              18.470641   1  97
3.0              17.389793   1  99
4.0              17.985849   1  99
5.0              14.412616   1  99
6.0              16.515856   1  99
7.0              17.413529   1  99
8.0              16.050344   1  96
9.0              16.083870   1  99
10.0             21.290293   1  99
11.0             16.494495   1  99
12.0             16.089124   1  99
13.0             16.787994   1  99
14.0             16.896188   1  99
15.0             16.442397   1  99
16.0             17.812191   1  99
17.0             16.081831   1  99
18.0             19.198872   1  99
19.0             17.404819   1  99
20.0             16.670273   1  99
21.0             23.252093   1  98

In [9]:
# Creates a new column 'max_order' with the maximum 'order_number' for each 'user_id' using transform()  
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

# Display the 15 first rows of dataframe
ords_prods_merge.head(15)

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0          2       33120                  1          1   202279    prior   
1          2       28985                  2          1   202279    prior   
2          2        9327                  3          0   202279    prior   
3          2       45918                  4          1   202279    prior   
4          2       30035                  5          0   202279    prior   
5          2       17794                  6          1   202279    prior   
6          2       40141                  7          1   202279    prior   
7          2        1819                  8          1   202279    prior   
8          2       43668                  9          0   202279    prior   
9          3       33754                  1          1   205970    prior   
10         3       24838                  2          1   205970    prior   
11         3       17704                  3          1   205970    prior   
12         3       21903                  4          1   205970    prior   
13         3       17668                  5          1   205970    prior   
14         3       46667                  6          1   205970    prior   

    order_number  order_dow  order_hour_of_day  days_since_prior_order  ...  \
0              3          5                  9                     8.0  ...   
1              3          5                  9                     8.0  ...   
2              3          5                  9                     8.0  ...   
3              3          5                  9                     8.0  ...   
4              3          5                  9                     8.0  ...   
5              3          5                  9                     8.0  ...   
6              3          5                  9                     8.0  ...   
7              3          5                  9                     8.0  ...   
8              3          5                  9                     8.0  ...   
9             16          5                 17                    12.0  ...   
10            16          5                 17                    12.0  ...   
11            16          5                 17                    12.0  ...   
12            16          5                 17                    12.0  ...   
13            16          5                 17                    12.0  ...   
14            16          5                 17                    12.0  ...   

                                         product_name  aisle_id department_id  \
0                                  Organic Egg Whites      86.0          16.0   
1                               Michigan Organic Kale      83.0           4.0   
2                                       Garlic Powder     104.0          13.0   
3                                      Coconut Butter      19.0          13.0   
4                                   Natural Sweetener      17.0          13.0   
5                                             Carrots      83.0           4.0   
6                    Original Unflavored Gelatine Mix     105.0          13.0   
7            All Natural No Stir Creamy Almond Butter      88.0          13.0   
8                             Classic Blend Cole Slaw     123.0           4.0   
9   Total 2% with Strawberry Lowfat Greek Strained...     120.0          16.0   
10                             Unsweetened Almondmilk      91.0          16.0   
11                                             Lemons     123.0           4.0   
12                               Organic Baby Spinach     123.0           4.0   
13    Unsweetened Chocolate Almond Breeze Almond Milk      91.0          16.0   
14                                Organic Ginger Root      83.0           4.0   

    prices     busiest day    busiest_days    Busiest_days     busiest_day  \
0     11.3  Regularly busy  Regularly busy  Regularly busy  Regularly busy   
1     13.4  Regularly busy  Regularly busy  Regularly busy  Regularly

In [11]:
# Deleting the two duplicate 'busiest_days' columns that were created due to an earlier error.
ords_prods_merge = ords_prods_merge.drop('busiest_days', axis=1)
ords_prods_merge = ords_prods_merge.drop('Busiest_days', axis=1)

In [13]:
# Renaming the 'busiest_day' column to 'busiest_days' in the dataframe
ords_prods_merge = ords_prods_merge.rename(columns={'busiest_day': 'busiest_days'})
# Display the 5 first rows of dataframe
ords_prods_merge.head()

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0         2       33120                  1          1   202279    prior   
1         2       28985                  2          1   202279    prior   
2         2        9327                  3          0   202279    prior   
3         2       45918                  4          1   202279    prior   
4         2       30035                  5          0   202279    prior   

   order_number  order_dow  order_hour_of_day  days_since_prior_order _merge  \
0             3          5                  9                     8.0   both   
1             3          5                  9                     8.0   both   
2             3          5                  9                     8.0   both   
3             3          5                  9                     8.0   both   
4             3          5                  9                     8.0   both   

   Unnamed: 0           product_name  aisle_id  department_id  prices  \
0     33122.0     Organic Egg Whites      86.0           16.0    11.3   
1     28987.0  Michigan Organic Kale      83.0            4.0    13.4   
2      9327.0          Garlic Powder     104.0           13.0     3.6   
3     45922.0         Coconut Butter      19.0           13.0     8.4   
4     30037.0      Natural Sweetener      17.0           13.0    13.7   

      busiest day    busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regularly busy        Average orders          8  
1  Regularly busy  Regularly busy        Average orders          8  
2  Regularly busy  Regularly busy        Average orders          8  
3  Regularly busy  Regularly busy        Average orders          8  
4  Regularly busy  Regularly busy        Average orders          8

In [15]:
# Assigning customer loyalty status based on 'max_order' values: 'Loyal customer' for max_order > 40, 'Regular customer' 
# for max_order between 10 and 40, and 'New customer' for max_order <= 10
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
# Printing the frequency of values in the 'loyalty_flag' column
ords_prods_merge['loyalty_flag'].value_counts()

loyalty_flag
Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: count, dtype: int64

In [19]:
# Displaying the first 30 unique 'user_id', 'loyalty_flag', and 'order_number' rows, without repeating 'user_id'
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].drop_duplicates(subset='user_id').head(30)

user_id      loyalty_flag  order_number
0     202279      New customer             3
9     205970  Regular customer            16
17    178520    Loyal customer            36
30    156122    Loyal customer            42
56     22352      New customer             4
59    142903  Regular customer            11
61      3107  Regular customer             5
62    139016  Regular customer            14
77    135442  Regular customer             4
92    143742  Regular customer             4
97    152610  Regular customer            22
112    45082      New customer             2
125    18194    Loyal customer            49
136    54901    Loyal customer            51
141   174840  Regular customer            18
144   118860    Loyal customer             3
172    83009  Regular customer             7
175   182912  Regular customer             1
183    88772  Regular customer            13
188    30822  Regular customer            25
202    68572  Regular customer            21
216   193635  Regular customer            19
219    59897  Regular customer            19
233   153404  Regular customer             2
241   129389  Regular customer            16
268    98256    Loyal customer            29
284    81606  Regular customer            14
289   194309  Regular customer             2
292   201744  Regular customer             7
302   119216      New customer             4

# Step 2

In [21]:
# Groups the entire dataframe by 'department_id' and calculates the mean of 'order_number' for each group  
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.665606
17.0             15.694469
18.0             19.310397
19.0             17.177343
20.0             16.473447
21.0             22.902379

# Step 3

### Analysis of 'order_number' Means by 'department_id'

- **Entire Dataframe Results:**
  - The overall means for most departments are slightly lower compared to the subset.
  - Departments such as **bulk (ID: 10)** and **missing (ID: 21)** show higher mean values in both the entire dataset and the subset.

- **Subset Results:**
  - Slightly higher means for most departments, indicating that the subset likely contains a higher proportion of frequent shoppers.

#### Observations:
- **Subset bias:** The subset appears to represent a more engaged group of customers (frequent purchasers).
- **Key Differences:**
  - **Bulk (ID: 10)** shows the largest differences in means (21.29 in the subset vs. 20.20 overall).
  - The **missing department (ID: 21)** has the highest mean order number, consistently above 22.
- These differences suggest that customer behavior varies depending on the subset

# Step 4

In [23]:
# Creating a 'loyalty_flag' based on the max order number for each user
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

# Assigning loyalty categories using loc()
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

# Checking the distribution of loyalty categories
ords_prods_merge['loyalty_flag'].value_counts()

loyalty_flag
Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: count, dtype: int64

# Step 5

In [25]:
# Checking basic statistics of product prices for each loyalty category
loyalty_price_stats = ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

# Displaying the results
loyalty_price_stats

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386336  328.017787  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294670  597.560299  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495717  539.720919  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0

### **Spending Habits by Customer Loyalty Group**

- **Loyal Customers** → Spend the least (**10.39 avg. price**), likely more price-conscious.  
- **Regular Customers** → Moderate spending (**12.50 avg. price**), balancing budget and premium items.  
- **New Customers** → Spend the most (**13.29 avg. price**), possibly exploring higher-priced products.  

**Observations:**  
- **Median price (~7.40)** is similar across groups, meaning most purchases fall in the same range.  
- **High standard deviation** suggests varied spending behavior.  
- **Max price (99,999)** is an outlier and should be reviewed.  


# Step 6

In [27]:
# Calculate the average price for each user
avg_price_per_user = ords_prods_merge.groupby('user_id')['prices'].mean()

In [29]:
# Create a new column 'spending_flag' using loc based on the average price
ords_prods_merge['spending_flag'] = 'Low spender'  # Default to 'Low spender'
ords_prods_merge.loc[ords_prods_merge['user_id'].isin(avg_price_per_user[avg_price_per_user >= 10].index), 'spending_flag'] = 'High spender'
# Checking the distribution
ords_prods_merge['spending_flag'].value_counts()

spending_flag
Low spender     31800381
High spender      634678
Name: count, dtype: int64

# Step 7

In [31]:
# Calculate the median of 'days_since_prior_order' for each user
median_days_per_user = ords_prods_merge.groupby('user_id')['days_since_prior_order'].median()

In [33]:
# Create a new column 'order_frequency_flag' using loc based on the median value
ords_prods_merge['order_frequency_flag'] = 'Non-frequent customer'  # Default to 'Non-frequent customer'
# Update the flag based on the median value
ords_prods_merge.loc[ords_prods_merge['user_id'].isin(median_days_per_user[median_days_per_user <= 10].index), 'order_frequency_flag'] = 'Frequent customer'
ords_prods_merge.loc[ords_prods_merge['user_id'].isin(median_days_per_user[(median_days_per_user > 10) & (median_days_per_user <= 20)].index), 'order_frequency_flag'] = 'Regular customer'
# Checking the distribution
ords_prods_merge['order_frequency_flag'].value_counts()

order_frequency_flag
Frequent customer        21578273
Regular customer          7217066
Non-frequent customer     3639720
Name: count, dtype: int64

# Step 9

In [35]:
# Display the dataframe
ords_prods_merge.head()

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0         2       33120                  1          1   202279    prior   
1         2       28985                  2          1   202279    prior   
2         2        9327                  3          0   202279    prior   
3         2       45918                  4          1   202279    prior   
4         2       30035                  5          0   202279    prior   

   order_number  order_dow  order_hour_of_day  days_since_prior_order  ...  \
0             3          5                  9                     8.0  ...   
1             3          5                  9                     8.0  ...   
2             3          5                  9                     8.0  ...   
3             3          5                  9                     8.0  ...   
4             3          5                  9                     8.0  ...   

  aisle_id  department_id prices     busiest day    busiest_days  \
0     86.0           16.0   11.3  Regularly busy  Regularly busy   
1     83.0            4.0   13.4  Regularly busy  Regularly busy   
2    104.0           13.0    3.6  Regularly busy  Regularly busy   
3     19.0           13.0    8.4  Regularly busy  Regularly busy   
4     17.0           13.0   13.7  Regularly busy  Regularly busy   

   busiest_period_of_day max_order  loyalty_flag spending_flag  \
0         Average orders         8  New customer   Low spender   
1         Average orders         8  New customer   Low spender   
2         Average orders         8  New customer   Low spender   
3         Average orders         8  New customer   Low spender   
4         Average orders         8  New customer   Low spender   

    order_frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 23 columns]

In [37]:
# Export data ords_prods_merge dataframe with the new columns to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_48.pkl'))